### ISSUE TITLE
Alternative Direct3D-Compositor synchronization techniques

### ISSUE THREAD

**Issue by jh-go** [bigfatbrowncat/noflicker_directx_window#1]
Hello @bigfatbrowncat, thank you for providing this repository. Currently, two consecutive `IDXGISwapChain::Present()` calls are used to ensure that the compositor (DWM) uses a correctly-sized Direct3D frame. Why this works in the first place seems to be a bit of a mystery though, so I'd like to point out a few alternatives:

1. [`DWMFlush()`](https://learn.microsoft.com/en-us/windows/win32/api/dwmapi/nf-dwmapi-dwmflush) is available since Windows Vista but probably can't cope with some edge cases (e.g., if a window is stretched over two displays with different refresh rates)
2. [`DCompositionWaitForCompositorClock()`](https://learn.microsoft.com/en-us/windows/win32/api/dcomp/nf-dcomp-dcompositionwaitforcompositorclock) promises to solve the shortcomings `DWMFlush()` but is only supported on Windows 11 21H2 (Build 22000) and newer

While it is surely advisable to work *with* the compositor, we could also skip it and wait for vertical blank directly:

3. [`D3DKMTWaitForVerticalBlankEvent()`](https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/d3dkmthk/nf-d3dkmthk-d3dkmtwaitforverticalblankevent) is the most accurate way to wait for vertical blank, since this function is directly provided by the Direct3D driver
4. [`IDXGIOutput::WaitForVBlank()`](https://learn.microsoft.com/en-us/windows/win32/api/dxgi/nf-dxgi-idxgioutput-waitforvblank) seems to be a `IDXGIOutput`-specific version of `D3DKMTWaitForVerticalBlankEvent()`

In the case of the latter two options, we would have to figure out on which `HMONITOR`/`IDXGIOutput` our window is displayed.

Further information:
https://stackoverflow.com/questions/49244480/correct-way-to-wait-for-vblank-on-windows-10-in-windowed-mode
https://issues.chromium.org/issues/40409726
https://armageddongames.net/archive/index.php/t-96793.html

**Comment by wangwenx190** [comment]
@jh-go Hi, thanks for your great ideas! But I tried your solutions listed above, none of them can really get rid of the resize flicker.

DwmFlush: can reduce the flicker, but can't totally remove flicker.
DCompositionWaitForCompositorClock: similar to DwmFlush, can reduce but can't totally remove flicker.
IDXGIOutput::WaitForVBlank: similar to DwmFlush as well.

I even tried to combine DwmFlush + DCompositionWaitForCompositorClock/IDXGIOutput::WaitForVBlank, or disable V-Sync and render as fast as possible, the results vary a lot but none of them can totally remove the flicker.

In the end there still doesn't exist a perfect solution that really solves the resize flicker. But the UWP applications also use D3D/DComp however they don't flicker at all no matter how you resize the window, I'm so curious how MS achieve this.

By the way, I'm testing on a 4K 240Hz monitor, with GSync turned on, rendering through NVIDIA 4090 with latest driver on latest Win11.

**Comment by wangwenx190** [comment]
I may have to mention, that this repo also flickers on my PC, so it seems the magical trick that this repo uses doesn't work for everyone.

**Comment by bigfatbrowncat** [comment]
> I may have to mention, that this repo also flickers on my PC, so it seems the magical trick that this repo uses doesn't work for everyone.

Very strange... I haven't tried with GSync, but on my NVidia GPU it worked like a charm. Could you please try with 60FPS without GSync? I wonder if it works in that case...

**Comment by wangwenx190** [comment]
> > I may have to mention, that this repo also flickers on my PC, so it seems the magical trick that this repo uses doesn't work for everyone.
> 
> Very strange... I haven't tried with GSync, but on my NVidia GPU it worked like a charm. Could you please try with 60FPS without GSync? I wonder if it works in that case...

It seems there's much magic behind the scene so actually I'm not surprised.

For my own project, after I add a DwmFlush() call in WM_NCCALCSIZE, the flicker totally disappeared, yeah, totally gone. But I was testing on a 1080p 60Hz monitor with NVIDIA 3060. After I copied the same binary to my another PC with the hardware I mentioned above (4K 240Hz etc ...), the flicker appears again. And I also sent the same binary to my friend to help me test, he run the executable on 1080p 240Hz with NVIDIA 2070, he reported that window resize flicker a lot. So, it seems that current solutions only work for some specific cases.

**Comment by wangwenx190** [comment]
I think whether GSync is enabled or not or whether the monitor is 60Hz or not is not the most critical thing. Look at UWP applications, no matter how I change NVIDIA settings, no matter display on what monitor, they always behave perfectly, so there must be an unknown solution that can fix the root cause of the resize flicker, not just some black magic workarounds.

**Comment by apkipa** [comment]
> In the end there still doesn't exist a perfect solution that really solves the resize flicker. But the UWP applications also use D3D/DComp however they don't flicker at all no matter how you resize the window, I'm so curious how MS achieve this.

I'm afraid that there really isn't a way for D3D-based applications to properly synchronize with compositor (DWM), although it works much better if you use the old BitBlt model instead of flip model. And yes, while UWP (only XAML ones; Direct3D ones still flicker) never flickers, the techniques behind cannot be applied here because UWP never calls `IDXGISwapChain::Present()`; it just calls `IDCompositionDevice::Commit()` to commit the entire visual tree to DComp. The final presentation is handled by DWM atomically. Then, UWP solves the flickering by:

* Marking the window as eligible for resize synchronization. This will let OS create a DComp synchronization handle during WM_SIZE, which can be retrieved by calling `GetResizeDCompositionSynchronizationObject()`.
* Making changes to the DComp visual tree, then calling the undocumented API `SynchronizedCommit(resizeHandle)` on the DComp device. Calling this (instead of `Commit()`) is necessary, or the changes will be visible **before** the window frame redraws (flickers in a reverse way).
* Calling `Windows::UI::Core::CoreWindowResizeManager::GetForCurrentView().NotifyLayoutCompleted()`. This tells DWM to finally start composing a new frame for the window.

It cannot be done for D3D since AFAIK it doesn't have the `SynchronizedCommit()` equivalent.

**Comment by bigfatbrowncat** [comment]
> It cannot be done for D3D since AFAIK it doesn't have the SynchronizedCommit() equivalent.

In my code I am using D3D window INSIDE the DComp tree. So this approach would work here!

Could you please try to make an MR with some proof of concept? Or at least send me a link to their code. I want to look how it is done.

**Comment by wangwenx190** [comment]
> Could you please try to make an MR with some proof of concept? Or at least send me a link to their code. I want to look how it is done.

Yeah, using D3D + DComp is common practice, I'm also curious what the code look like :)

**Comment by apkipa** [comment]
> In my code I am using D3D window INSIDE the DComp tree. So this approach would work here!

Well, I already tried this with zero success. Once you call `Commit()` to make the swapchain onscreen, its content updates are no longer coupled with DComp, so no synchronization unfortunately.

Anyway, if you are adventurous, here is my attempt. Feel free to adapt and see it yourself.
https://gist.github.com/apkipa/20cae438aef2a8633f99e10e0b90b11e
